pkg/simulation/simulator/user/user_sim.go

```diff
@@ -0,0 +1,262 @@
+	if su.Behavior.SendRegularMsg() {
+		reg_target := su.User.RegularContactList[su.Behavior.GetRandomizer().Intn(len(su.User.RegularContactList))]
+		reg_msg := su.makeRegularMessage(reg_target)
+		msgs = append(msgs, reg_msg)
+	}
+
+	return msgs
+}
+
+func (su *SimulatedUser) SendMessage(msg Types.Msg) {
+	if su == nil {
+		return
+	}
+
+	su.logger <- Types.MsgEvent{Msg: msg, EventType: "Send"}
+
+	// Ensures messages are sent to the docker container at a rate it can handle
+	su.GlobalLock.Lock()
+	su.mu.Lock()
+	su.Process.Cmd([]byte(fmt.Sprintf("%v\n", msg.MsgContent)))
+	su.mu.Unlock()
+	time.Sleep(50 * time.Millisecond) //Important to avoid fucking processes, if I knew why I would fix the root problem
+	su.GlobalLock.Unlock()
+}
+
+func (su *SimulatedUser) OnReceive(msg Types.Msg) {
+	if su == nil {
+		return
+	}
+
+	su.logger <- Types.MsgEvent{EventType: "Receive", Msg: msg}
+
+	//Determine if Alice responds to the message
+	if !su.Behavior.WillRespond() {
+		return
+	}
+
+	//TODO: Determine whether the message is regular or the entirety of a deniable message has been received
+
+	res := su.makeRegularMessage(msg.From)
+	go su.SendMessage(res)
+}
+
+func (su *SimulatedUser) MessageListener() {
+	for {
+		select {
+		case <-su.stopChan:
+			break
+		default:
+			time.Sleep(time.Duration(readTimeout * time.Second))
+			su.mu.Lock()
+			su.Process.Cmd([]byte("read\n"))
+			su.mu.Unlock()
+			for su.Process.Buffer.Len() != 0 {
+				line, err := su.Process.Buffer.ReadString(byte('\n'))
+				if len(line) == 0 {
+					fmt.Println("Encountered error: %v \n", err)
+					break
+				}
+
+				if err != nil {
+					break
+				}
+
+				if len(line) == 1 {
+					continue
+				}
+
+				splits := strings.Split(line, ":")
+				if splits[0] == "" || splits[0] == "\n" {
+					continue
+				}
+
+				sender := splits[0]
+				if len(sender) < 8 {
+					continue
+				}
+
+				sender = sender[8:len(sender)]
+				if sender == "\n" {
+					continue
+				}
+				for _, name := range su.User.RegularContactList {
+					if sender == name {
+						msg := Types.Msg{To: su.User.Nickname, From: name, MsgContent: splits[1], IsDeniable: false}
+						go su.OnReceive(msg)
+						break
+					}
+				}
+
+				for _, name := range su.User.DeniableContactList {
+					if sender == name {
+						msg := Types.Msg{To: su.User.Nickname, From: name, MsgContent: splits[1], IsDeniable: true}
+						go su.OnReceive(msg)
+						break
+					}
+				}
+			}
+		}
+	}
+}
+
+func (su *SimulatedUser) SetDeniableContacts(contacts []string) {
+	su.User.DeniableContactList = append(su.User.DeniableContactList, contacts...)
+}
+
+// Modifies all passed users' contact lists and ensures each user has a minimum number of contacts
+// in the range of min_contact_count and max_contact_count
+func CreateCommunicationNetwork(users *[]Types.SimUser, min_contact_count, max_contact_count int, r *rand.Rand) *[]Types.SimUser {
```

Comment:
Should be:
`func CreateCommunicationNetwork(users []Types.SimUser, min_contact_count, max_contact_count int, r *rand.Rand) []Types.SimUser
`